Modules/vc-module-export/VirtoCommerce.ExportModule.Data/Services/CsvExportProvider.cs

```diff
@@ -2,123 +2,160 @@
                 {
-                    if (!cacheGetMethods.ContainsKey(memberInfo))
+                    var propertyName = propertyNames[i];
+                    var propertyInfo = currentType.GetProperty(propertyName, BindingFlags.Public | BindingFlags.Instance);
+
+                    if (propertyInfo?.CanRead != true)
+                    {
+                        break;
+                    }
+
+                    // Do not write enumerables (ICollection, etc.)
+                    if (IsEnumerableEntityProperty(propertyInfo))
+                    {
+                        break;
+                    }
+
+                    // Add memberMap
+                    if (i == propertyNames.Length - 1)
+                    {
+                        var memberMap = CreateMemberMap(currentType, propertyInfo, includedPropertyInfo.Name, ref columnIndex);
+
+                        currentClassMap.MemberMaps.Add(memberMap);
+                        currentClassMap = this;
+                    }
+                    // Working with nested properties - create or get References
+                    else
                     {
-                        var getMethod = DynamicMethodGenerator.CreateDynamicMethod(obj.GetType(), (PropertyInfo)memberInfo);
-                        cacheGetMethods.Add(memberInfo, getMethod);
+                        var referenceMap = currentClassMap.ReferenceMaps.Find(propertyInfo);
+                        currentType = propertyInfo.PropertyType;
+
+                        if (referenceMap == null)
+                        {
+                            var referenceClassMapType = typeof(EmptyClassMapImpl<>).MakeGenericType(new[] { currentType });
+
+                            referenceMap = new MemberReferenceMap(propertyInfo, (ClassMap)Activator.CreateInstance(referenceClassMapType));
+                            currentClassMap.ReferenceMaps.Add(referenceMap);
+                        }
+
+                        currentClassMap = referenceMap.Data.Mapping;
                     }
                 }
             }
-            return cacheGetMethods[memberInfo](obj);
         }
 
-        public void WriteMetadata(ExportedTypeMetadata metadata)
+        private static bool IsEnumerableEntityProperty(PropertyInfo propertyInfo)
         {
-            var firstProperty = true;
-            foreach (var propertyInfo in metadata.PropertiesInfo)
-            {
-                if (!firstProperty) _streamWriter.Write(";");
-                _streamWriter.Write(@"""");
-                _streamWriter.Write(string.IsNullOrEmpty(propertyInfo.ExportName) ? propertyInfo.Name : propertyInfo.ExportName);
-                _streamWriter.Write(@"""");
-                firstProperty = false;
-            }
-            _streamWriter.WriteLine();
+            var type = propertyInfo.PropertyType;
+            return type.GetInterfaces().Any(x => x.IsGenericType && x.GetGenericTypeDefinition() == typeof(IEnumerable<>))
```

Comment:
Put each condition on a separate line.
Multi-line blocks of code should be separated with empty line from other code.